be/src/storage/rowset_column_update_state.cpp

```diff
@@ -0,0 +1,524 @@
+// limitations under the License.
+
+#include "rowset_column_update_state.h"
+
+#include "common/tracer.h"
+#include "fs/fs_util.h"
+#include "gutil/strings/substitute.h"
+#include "serde/column_array_serde.h"
+#include "storage/chunk_helper.h"
+#include "storage/delta_column_group.h"
+#include "storage/primary_key_encoder.h"
+#include "storage/rowset/column_iterator.h"
+#include "storage/rowset/rowset.h"
+#include "storage/rowset/rowset_options.h"
+#include "storage/rowset/segment_options.h"
+#include "storage/rowset/segment_rewriter.h"
+#include "storage/tablet.h"
+#include "storage/tablet_meta_manager.h"
+#include "storage/update_manager.h"
+#include "util/defer_op.h"
+#include "util/phmap/phmap.h"
+#include "util/stack_util.h"
+#include "util/time.h"
+
+namespace starrocks {
+
+RowsetColumnUpdateState::RowsetColumnUpdateState() = default;
+
+RowsetColumnUpdateState::~RowsetColumnUpdateState() {
+    if (!_status.ok()) {
+        LOG(WARNING) << "bad RowsetColumnUpdateState released tablet:" << _tablet_id;
+    }
+}
+
+Status RowsetColumnUpdateState::load(Tablet* tablet, Rowset* rowset) {
+    if (UNLIKELY(!_status.ok())) {
+        return _status;
+    }
+    std::call_once(_load_once_flag, [&] {
+        _tablet_id = tablet->tablet_id();
+        _status = _do_load(tablet, rowset);
+        if (!_status.ok()) {
+            LOG(WARNING) << "load RowsetColumnUpdateState error: " << _status << " tablet:" << _tablet_id << " stack:\n"
+                         << get_stack_trace();
+            if (_status.is_mem_limit_exceeded()) {
+                LOG(WARNING) << CurrentThread::mem_tracker()->debug_string();
+            }
+        }
+    });
+    return _status;
+}
+
+Status RowsetColumnUpdateState::_load_upserts(Rowset* rowset, uint32_t idx) {
+    RowsetReleaseGuard guard(rowset->shared_from_this());
+    if (_upserts.size() == 0) {
+        _upserts.resize(rowset->num_update_files());
+        _update_chunk_cache.resize(rowset->num_update_files());
+    }
+    if (_upserts.size() == 0 || _upserts[idx] != nullptr) {
+        return Status::OK();
+    }
+
+    const int64_t DEFAULT_CONTAINER_SIZE = 4096;
+    OlapReaderStatistics stats;
+    auto& schema = rowset->schema();
+    vector<uint32_t> pk_columns;
+    for (size_t i = 0; i < schema.num_key_columns(); i++) {
+        pk_columns.push_back((uint32_t)i);
+    }
+    std::vector<uint32_t> update_columns;
+    const auto& txn_meta = rowset->rowset_meta()->get_meta_pb().txn_meta();
+    for (uint32_t cid : txn_meta.partial_update_column_ids()) {
+        update_columns.push_back(cid);
+    }
+    Schema pkey_schema = ChunkHelper::convert_schema(schema, pk_columns);
+    Schema update_schema = ChunkHelper::convert_schema(schema, update_columns);
+    std::unique_ptr<Column> pk_column;
+    if (!PrimaryKeyEncoder::create_column(pkey_schema, &pk_column).ok()) {
+        CHECK(false) << "create column for primary key encoder failed";
```

Comment:
Don't use CHECK it will cause process crash, change it to DCHECK and return Status::IternalError

Reply:
You are right, I use `DCHECK` to quick find out bug when write and test the new code, they should removed in office code.